pex/resolve/lockfile/subset.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import os
+from collections import OrderedDict
+
+from pex.common import pluralize
+from pex.network_configuration import NetworkConfiguration
+from pex.orderedset import OrderedSet
+from pex.requirements import LocalProjectRequirement, parse_requirement_strings
+from pex.resolve.locked_resolve import DownloadableArtifact, Resolved
+from pex.resolve.lockfile.model import Lockfile
+from pex.resolve.requirement_configuration import RequirementConfiguration
+from pex.result import Error
+from pex.targets import Target, Targets
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Dict, Iterable, Optional, Tuple, Union
+
+    import attr  # vendor:skip
+
+    from pex.requirements import ParsedRequirement
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class Subset(object):
+    target = attr.ib()  # type: Target
+    resolved = attr.ib()  # type: Resolved
+
+
+@attr.s(frozen=True)
+class SubsetResult(object):
+    requirements = attr.ib()  # type: Tuple[ParsedRequirement, ...]
+    subsets = attr.ib()  # type: Tuple[Subset, ...]
+
+
+def subset(
+    targets,  # type: Targets
+    lock,  # type: Lockfile
+    requirement_configuration=RequirementConfiguration(),  # type: RequirementConfiguration
+    network_configuration=None,  # type: Optional[NetworkConfiguration]
+    build=True,  # type: bool
+    use_wheel=True,  # type: bool
+    prefer_older_binary=False,  # type: bool
+    transitive=True,  # type: bool
+    include_all_matches=False,  # type: bool
+):
+    # type: (...) -> Union[SubsetResult, Error]
+
+    with TRACER.timed("Parsing requirements"):
```

Comment:
This was extracted from the lock_resolver here: https://github.com/pantsbuild/pex/pull/1831/files#diff-aad7222a03ff8182eb9c406092214ea24c7811b23dbdb9f47bfedc85aab327d0L240